src/main/java/biz/paluch/logging/gelf/intern/GelfMessage.java

```diff
@@ -402,8 +403,42 @@ public ByteBuffer toTCPBuffer(ByteBuffer buffer) {
         return slices;
     }
 
-    public int getCurrentMillis() {
-        return (int) System.currentTimeMillis();
+    byte[] generateMsgId() {
+        // Considerations about generating the message ID: The GELF documentation suggests to
+        // "[g]enerate [the id] from millisecond timestamp + hostname for example":
+        // https://docs.graylog.org/en/3.1/pages/gelf.html#chunking
+        //
+        // However, relying on current time in milliseconds on the same system will result in a high
+        // collision probability if lots of messages are generated quickly. Things will be even
+        // worse if multiple servers send to the same log server. Adding the hostname is not
+        // guaranteed to help, and if the hostname is the FQDN it is even unlikely to be unique at
+        // all.
+        //
+        // The GELF module used by Logstash uses the first eight bytes of an MD5 hash of the current
+        // time as floating point, a hyphen, and an eight byte random number:
+        // https://github.com/logstash-plugins/logstash-output-gelf
+        // https://github.com/graylog-labs/gelf-rb/blob/master/lib/gelf/notifier.rb#L239 It probably
+        // doesn't have to be that clever:
+        //
+        // Using the timestamp plus a random number will mean we only have to worry about collision
+        // of random numbers within the same milliseconds. How short can the timestamp be before it
+        // will collide with old timestamps? Every second Graylog will evict expired messaged (5
+        // seconds old) from the pool:
+        // https://github.com/Graylog2/graylog2-server/blob/master/graylog2-server/src/main/java/org/graylog2/inputs/codecs/GelfChunkAggregator.java
+        // Thus, we just need six seconds which will require two bytes. Then we can spend six bytes
+        // on a random number.
+
+        return ByteBuffer.allocate(8).putLong(getRandomLong())
+                // Overwrite the last two bytes with the timestamp.
+                .putShort(6, getCurrentTimeMillis()).array();
+    }
+
+    long getRandomLong() {
+        return rand.nextLong();
+    }
+
+    short getCurrentTimeMillis() {
+        return (short) System.currentTimeMillis();
```

Comment:
How about using proper bitmasking to make the usage of the two last timestamp bytes more obvious?

Reply:
You're right. I even realize that I can just return longs and use bitmasking all the way instead of fiddling with ByteBuffer.